### PR TITLE
[move] Initial doc migration for 2024

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

NOTE: WIP so not all the docs are done, just trying to give a early heads up, mostly for @ronny-mysten since its gonna be large 

- First stab at rewriting the docs for Move 2024
- I have not really documented anything "new" mostly verbiage updating, removing scripts, putting `public` on structs, etc
- More to come

## Test Plan 

- 👀 

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: external-crates/move/crates/move-compiler-transactional-tests/tests/constants/const_in_const.move [hunk 1]
```diff
@@ -10,6 +10,7 @@ const C_TWO: u64 = 2;
 const C_VEC: vector<u64> = vector[C_ZERO, C_ONE, C_TWO];
 const C_FIVE: u64 = 5;
 const C_V2: vector<u64> = vector[C_ZERO, C_FIVE];
+const C_VS: vector<vector<u64>> = vector[C_VEC, C_V2];
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler-transactional-tests/tests/constants/const_in_const.move:13]: Added this to double check myself while writing :)

File: external-crates/move/documentation/book/TODO [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+- Labels
+- Break with value
+- nested use
+- syntax annotations
+
+Tim
+
+- postfix abilities
+- positional fields
+- package system
```
Comment (tnowacki) [inline external-crates/move/documentation/book/TODO:30]: Hope y'all don't mind @cgswords @tzakian

File: external-crates/move/documentation/book/TODO [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+- Methods
+- unit tests abort codes
+
+Cameron
+
+- Labels
+- Break with value
+- nested use
+- syntax annotations
```
Comment (tnowacki) [inline external-crates/move/documentation/book/TODO:24]: I don't think we have an attributes section... so not sure what to say here

**Review by amnn — COMMENTED** [review]
Happy to take a look at things once the content is there, but don't want to review stuff while it's still in motion, so feel free to re-request review from me when ready.

File: external-crates/move/documentation/book/src/integers.md [hunk 4]
```diff
@@ -66,78 +71,94 @@ let hex_u256: u256 = 0x1123_456A_BCDE_F;
...
+| `>`    | greater than             |
+| `<=`   | less than or equal to    |
+| `>=`   | greater than or equal to |
 
 ### Equality
 
-Like all types with [`drop`](./abilities.md) in Move, all integer types support the ["equal"](./equality.md) and ["not equal"](./equality.md)  operations. Both arguments need to be of the same type. If you need to compare integers of different types, you will need to [cast](#casting) one of them first.
+Like all types with [`drop`](./abilities.md) in Move, all integer types support the
+["equal"](./equality.md) and ["not equal"](./equality.md) operations. Both arguments need to be of
+the same type. If you need to compare integers of different types, you will need to [cast](#casting)
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/integers.md:137]:
```suggestion
the same type. If you need to compare integers of different types, you must [cast](#casting)
```

File: external-crates/move/documentation/book/src/integers.md [hunk 5]
```diff
@@ -66,78 +71,94 @@ let hex_u256: u256 = 0x1123_456A_BCDE_F;
...
-| Syntax     | Operation                                                                       | Aborts if
-|------------|---------------------------------------------------------------------------------|---------------------------------------
-| `(e as T)`| Cast integer expression `e` into an integer type `T` | `e` is too large to represent as a `T`
+| Syntax     | Operation                                            | Aborts if                              |
+| ---------- | ---------------------------------------------------- | -------------------------------------- |
+| `(e as T)` | Cast integer expression `e` into an integer type `T` | `e` is too large to represent as a `T` |
 
-Here, the type of `e` must be `8`, `16`, `32`, `64`, `128` or `256` and `T` must be `u8`, `u16`, `u32`, `u64`, `u128` oe `u256`.
+Here, the type of `e` must be `8`, `16`, `32`, `64`, `128` or `256` and `T` must be `u8`, `u16`,
+`u32`, `u64`, `u128` oe `u256`.
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/integers.md:161]:
```suggestion
`u32`, `u64`, `u128`, or `u256`.
```

File: external-crates/move/documentation/book/src/loops.md [hunk 6]
```diff
@@ -4,14 +4,15 @@ Move offers two constructs for looping: `while` and `loop`.
...
 
-Here is an example of simple `while` loop that computes the sum of the numbers from `1` to `n`:
+Here is an example of simple `while` loop that computes the sum of the numbers from `0` to `n`:
 
 ```move
 fun sum(n: u64): u64 {
-    let sum = 0;
-    let i = 1;
+    let mut sum = 0;
+    let muy i = 1;
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/loops.md:15]: Esto a propósito? mut tal vez...

File: external-crates/move/documentation/book/src/modules.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+```move
+fun example() {
+    0xC0FFEE::m::foo(@0xC0FFEE);
+}
+```
+
+While at the source level these two different accesses are equivalent, it is a best practice to
+always use the named address and not the numerical value assigned to that address.
+
+Module names can start with letters `a` to `z` or letters `A` to `Z`. After the first character,
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/modules.md:88]:
```suggestion
Module names can start with a lowercase letter from `a` to `z` or an uppercase letter from `A` to `Z`. After the first character,
```
Adding suggested words might help clear any confusion of whether the repeat is intentional. Wouldn't be needed in next line.

File: external-crates/move/documentation/book/src/modules.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+module names can contain underscores `_`, letters `a` to `z`, letters `A` to `Z`, or digits `0` to
+`9`.
+
+```move
+module a::my_module {}
+module a::foo_bar_42 {}
+```
+
+Typically, module names start with an lowercase letter. A module named `my_module` should be stored
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/modules.md:97]:
```suggestion
Typically, module names start with a lowercase letter. A module named `my_module` should be stored
```

File: external-crates/move/documentation/book/src/modules.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+in a source file named `my_module.move`.
+
+## Members
+
+All members inside a `module` block can appear in any order. Fundamentally, a module is a collection
+of [`types`](./structs.md) and [`functions`](./functions.md). The [`use`](./uses.md) keyword is used
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/modules.md:103]:
```suggestion
of [`types`](./structs.md) and [`functions`](./functions.md). The [`use`](./uses.md) keyword
```

File: external-crates/move/documentation/book/src/modules.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+to refer to members from other modules. The [`const`](./constants.md) keyword defines constants that
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/modules.md:104]:
```suggestion
refers to members from other modules. The [`const`](./constants.md) keyword defines constants that
```

File: external-crates/move/documentation/book/src/modules.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+can be used in the functions of a module.
+
+[`friend`](./friends.md) are deprecated concept for specifying a list of trusted modules. The
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/modules.md:107]:
```suggestion
The [`friend`](./friends.md) syntax is a deprecated concept for specifying a list of trusted modules. The
```

File: external-crates/move/documentation/book/src/references.md [hunk 12]
```diff
@@ -134,59 +132,57 @@ also be used. This terminology is used in error messages to concisely indicate t
...
+  8 │         let mut x: &u64 = &0;
+    │                    ---- The type: '&u64'
     ·
   3 │     fun read_and_assign(store: &mut u64, new_value: &u64) {
     │                                -------- Is not a subtype of: '&mut u64'
     │
 ```
 
-The only other types currently that has subtyping are [tuples](./tuples.md)
+The only other types currently that have subtyping are [tuples](./tuples.md)
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/references.md:185]:
```suggestion
The only other types that currently have subtyping are [tuples](./tuples.md)
```

File: external-crates/move/documentation/book/src/references.md [hunk 13]
```diff
@@ -210,26 +206,20 @@ references before writes.
...
-
 One could imagine a fancier, more expressive, type system that would allow references to be stored
-in structs _and_ ban those structs from existing in global storage. We could perhaps allow
-references inside of structs that do not have the `store` [ability](./abilities.md), but that would
-not completely solve the problem: Move has a fairly complex system for tracking static reference
-safety, and this aspect of the type system would also have to be extended to support storing
-references inside of structs. In short, Move's type system (particularly the aspects around
-reference safety) would have to expand to support stored references. But it is something we are
-keeping an eye on as the language evolves.
+in structs. We could perhaps allow references inside of structs that do not have the `store`
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/references.md:218]:
```suggestion
in structs. We could allow references inside of structs that do not have the `store`
```

File: external-crates/move/documentation/book/src/references.md [hunk 14]
```diff
@@ -210,26 +206,20 @@ references before writes.
...
+[ability](./abilities.md), but that the core difficulty is that, Move has a fairly complex system
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/references.md:219]:
```suggestion
[ability](./abilities.md), but the core difficulty is that Move has a fairly complex system
```

File: external-crates/move/documentation/book/src/references.md [hunk 15]
```diff
@@ -210,26 +206,20 @@ references before writes.
...
+for tracking static reference safety, and this aspect of the type system would also have to be
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/references.md:220]:
```suggestion
for tracking static reference safety. This aspect of the type system would also have to be
```
As a noob, I'm not sure what "this aspect" is referring to here.

File: external-crates/move/documentation/book/src/tuples.md [hunk 16]
```diff
@@ -1,23 +1,25 @@
...
-in the return position for a function); they cannot be bound to local variables; they cannot be
-stored in structs; and tuple types cannot be used to instantiate generics.
-
-Similarly, [unit `()`](https://en.wikipedia.org/wiki/Unit_type) is a type created by the Move source language in order to be expression based.
-The unit value `()` does not result in any runtime value. We can consider unit`()` to be an empty
-tuple, and any restrictions that apply to tuples also apply to unit.
+[first-class value](https://en.wikipedia.org/wiki/First-class_citizen). However, in order to support
+multiple return values, Move has tuple-like expressions. These expressions do not result in a
+concrete value at runtime (there are no tuples in the bytecode), and as a result they are very
+limited: they can only appear in expressions (usually in the return position for a function); they
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/tuples.md:7]:
```suggestion
limited: 
- They can only appear in expressions (usually in the return position for a function).
- They cannot be bound to local variables.
```

File: external-crates/move/documentation/book/src/tuples.md [hunk 17]
```diff
@@ -1,23 +1,25 @@
...
+cannot be bound to local variables; they cannot be stored in structs; and tuple types cannot be used
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/tuples.md:8]:
```suggestion
- They cannot be bound to local variables.
- They cannot be stored in structs.
- Tuple types cannot be used to instantiate generics.
```

File: external-crates/move/documentation/book/src/tuples.md [hunk 18]
```diff
@@ -1,23 +1,25 @@
...
+to instantiate generics.
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/tuples.md:9]:
```suggestion

```

File: external-crates/move/documentation/book/src/tuples.md [hunk 19]
```diff
@@ -92,23 +91,23 @@ module example {
...
 ```
 
 For more details, see [Move Variables](./variables.md).
 
 ## Subtyping
 
-Along with references, tuples are the only types that have [subtyping](https://en.wikipedia.org/wiki/Subtyping) in Move. Tuples do have
-subtyping only in the sense that subtype with references (in a covariant way).
+Along with references, tuples are the only types that have
+[subtyping](https://en.wikipedia.org/wiki/Subtyping) in Move. Tuples do have subtyping only in the
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/tuples.md:109]:
```suggestion
[subtyping](https://en.wikipedia.org/wiki/Subtyping) in Move. Tuples have subtyping only in the
```

File: external-crates/move/documentation/book/src/unit-testing.md [hunk 20]
```diff
@@ -22,7 +28,11 @@ fun this_is_a_test() { ... }
 fun this_is_not_correct(arg: signer) { ... }
 ```
 
-A test can also be annotated as an `#[expected_failure]`. This annotation marks that the test should is expected to raise an error. You can ensure that a test is aborting with a specific abort code by annotating it with `#[expected_failure(abort_code = <code>)]`, if it then fails with a different abort code or with a non-abort error the test will fail. Only functions that have the `#[test]` annotation can also be annotated as an #`[expected_failure]`.
+A test can also be annotated as an `#[expected_failure]`. This annotation marks that the test should
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/unit-testing.md:31]:
```suggestion
A test can also be annotated as an `#[expected_failure]`. This annotation marks that the test
```

File: external-crates/move/documentation/book/src/vector.md [hunk 21]
```diff
@@ -149,21 +149,31 @@ fun destroy_droppable_vector<T: drop>(vec: vector<T>) {
...
-let y = copy x; // compiler error without the copy!
+let x = vector[10];
+let y = x; // implicit copy
+let z = x;
+(y, z)
 ```
 
-Copies of large vectors can be expensive, so the compiler requires explicit `copy`'s to make it
-easier to see where they are happening.
+Keep in mind, copies of large vectors can be expensive, If this is a concern, annotating the
```
Comment (ronny-mysten) [inline external-crates/move/documentation/book/src/vector.md:162]:
```suggestion
Keep in mind, copies of large vectors can be expensive. If this is a concern, annotating the
```

**Review by ronny-mysten — APPROVED** [review]
Apologies. I know many of the comments are not changes you made but I was there, so...

File: external-crates/move/documentation/book/src/equality.md [hunk 22]
```diff
@@ -88,6 +84,24 @@ i == s; // ERROR!
 //   ^ expected an argument of type '&u64'
 ```
 
+### Automatic Borrowing
+
+Starting in Move 2024 edition, the `==` and `!=` operators will automatically borrow their operands
+if one of the operands is a reference and the other is not. This means that the following code will
+work without any errors
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply external-crates/move/documentation/book/src/equality.md]: Maybe? Thoughts, @cgswords?

File: external-crates/move/documentation/book/src/friends.md [hunk 23]
```diff
@@ -1,101 +1,93 @@
...
-  module m { friend 0x42::nonexistent; // ERROR! }
-  //                ^^^^^^^^^^^^^^^^^ Unbound module '0x42::nonexistent'
-  }
+  module 0x42::m { friend 0x42::nonexistent; // ERROR! }
+  //                      ^^^^^^^^^^^^^^^^^ Unbound module '0x42::nonexistent'
   ```
 
-- Friend modules must be within the same account address. (Note: this is not a technical requirement but rather a policy decision which *may* be relaxed later.)
+- Friend modules must be within the same account address. (Note: this is not a technical requirement
+  but rather a policy decision which _may_ be relaxed later.)
```
Comment (tnowacki) [inline external-crates/move/documentation/book/src/friends.md:72]:
```suggestion
- Friend modules must be within the same account address.
```

File: external-crates/move/documentation/book/src/constants.md [hunk 24]
```diff
@@ -47,27 +32,28 @@ contain underscores `_`, letters `a` to `z`, letters `A` to `Z`, or digits `0` t
...
-const MY_ERROR_CODE: u64 = 0;
+const EMyErrorCode: u64 = 0;
 const ADDRESS_42: address = @0x42;
 ```
 
 Even though you can use letters `a` to `z` in a constant. The
 [general style guidelines](./coding-conventions.md) are to use just uppercase letters `A` to `Z`,
-with underscores `_` between each word.
+with underscores `_` between each word. For error codes, we use `E` as a prefix and then upper camel
+case for the rest of the name.
```
Comment (tnowacki) [inline external-crates/move/documentation/book/src/constants.md:42]:
```suggestion
with underscores `_` between each word. For error codes, we use `E` as a prefix and then upper camel
case (also known as Pascal case) for the rest of the name, as seen in `EMyErrorCode`.
```

**Pull request merged**